### PR TITLE
Store metadata internally in utf8 [WIP]

### PR DESCRIPTION
**Pull request by amangrewal**
Here's a first draft. It works for me, but I don't have an easy way of inserting invalid utf8 characters into metadata yet.

TODO

- More testing

Here's a list of open issues that this pull request *should* close:
#894, #885, #847, #605.

### CODE REVIEW

**Comment by nefthy** [comment]
You could do something like this to produce invalid utf-8:
```
t=`echo öäü | iconv -f utf8 -t latin1`
id3tag -a$t broken_artist.mp3
```

File: uchar.c [hunk 1]
```diff
@@ -520,14 +584,16 @@ int u_to_ascii(char *dst, const char *src, int len)
 	return i;
 }
 
-void u_to_utf8(char *dst, const char *src)
+const char *u_to_utf8(char *dst, const char *src)
```
Comment (nefthy) [inline uchar.c]: would you mind explaining this signature change? It is not obvious to me why it is needed.

Comment (geek1011) [reply uchar.c]: Looks like it is used to allow for chaining, which makes the code look a lot nicer.

Comment (amangrewal) [reply uchar.c]: Yeah, it's just for chaining. There's no need for it beyond aesthetics.

Comment (nefthy) [reply uchar.c]: I don't like this change, as it is purely aesthetic and chaining is not a widely used idiom in the cmus code base.

**Comment by nefthy** [comment]
Hm, I just stumbled across this in the documentation:
```
id3_default_charset (ISO-8859-1)

Default character set to use for ID3v1 and broken ID3v2 tags.
NOTE: This is used only if the tag is not valid UTF-8.
```
How does this option interact with this PR?

**Comment by amangrewal** [comment]
Good catch, I have to look into it.

From a cursory glance, the behavior depends on which input plugin is used. I'll provide an update when I have something more definite.

**Comment by amangrewal** [comment]
I've updated the above script. I was having issues with id3tag and certain file types.

I used ```grep id3 *.c *.h``` in ip to get a list of all file types/plugins that should support id3. The list of plugins is aac, mad, mp4, nomad, and wavpack. Using ```cmus --plugins``` to get the list of file types gives .aac, .m4a, .m4b, .mp2, .mp3, .mp4, and .wv.

For the file types that id3tag works with (aac, mp2, mp3, .wv), any invalid utf8 is treated as ISO-8859-1. When tagging with id3tag, .aac, .mp2, and .mp3 treat valid utf8 as ISO-8859-1; .wv treats it as utf8. When tagging with ffmpeg, .mp3 and .wv treat it as utf8; .aac and .mp2 don't seem to be tagged by ffmpeg.

In other words, embedded metadata is a complete mess.

Things I still need to test:
flac/cue
http streams
building without iconv

**Comment by nefthy** [comment]
I file with this id3 tag failes to parse. The file failed to parse before this patch, so I guess it's not a show stopper.

[meta.txt](https://github.com/cmus/cmus/files/3086419/meta.txt)

**Comment by nefthy** [comment]
I tested today by I adding about every mp3 file I have on my system. Everithing seems fine. I'll do some more deliberate tests these days. 

We should get some more testers on board before we merge this, since it is a major change. Do you think the patch is ready for testing/merging or are there open todos?

**Comment by amangrewal** [comment]
I tested the meta.txt file. It didn't cause a crash or a freeze, but perhaps I added the metadata differently.

I would like to have a few testers. There are no more open TODOs.

I couldn't manage to configure icecast properly so that cmus could play from it, so I wasn't able to test streams. Other than that, I've tested all the file types and configurations mentioned that I said I would test.

**Comment by nefthy** [comment]
@amangrewal it does not crush, it's just not added correctly. The filename is displayed instead of the meta data.

**Comment by nefthy** [comment]
Closing due to inactivity. Please feel free to reopen or create a new one.
